block/produce.go

```diff
@@ -36,51 +38,56 @@ func (m *Manager) ProduceBlockLoop(ctx context.Context) (err error) {
 	for {
 		select {
 		case <-ctx.Done():
-			return
+			return nil
 		case <-ticker.C:
 			// if empty blocks are configured to be enabled, and one is scheduled...
 			produceEmptyBlock := firstBlock || 0 == m.Conf.MaxIdleTime || nextEmptyBlock.Before(time.Now())
 			firstBlock = false
 
-			var (
-				block  *types.Block
-				commit *types.Commit
-			)
-			block, commit, err = m.ProduceAndGossipBlock(ctx, produceEmptyBlock)
+			block, commit, err := m.ProduceAndGossipBlock(ctx, produceEmptyBlock)
 			if errors.Is(err, context.Canceled) {
 				m.logger.Error("Produce and gossip: context canceled.", "error", err)
-				return
+				return nil
 			}
-			if errors.Is(err, types.ErrSkippedEmptyBlock) {
+			if errors.Is(err, types.ErrEmptyBlock) { // occurs if the block was empty but we don't want to produce one
 				continue
 			}
 			if errors.Is(err, ErrNonRecoverable) {
-				m.logger.Error("Produce and gossip: non-recoverable.", "error", err) // TODO: flush? or don't log at all?
 				uevent.MustPublish(ctx, m.Pubsub, &events.DataHealthStatus{Error: err}, events.HealthStatusList)
-				return
+				return err
 			}
 			if err != nil {
 				m.logger.Error("Produce and gossip: uncategorized, assuming recoverable.", "error", err)
 				continue
 			}
 
-			// If IBC transactions are present, set proof required to true
-			// This will set a shorter timer for the next block
-			// currently we set it for all txs as we don't have a way to determine if an IBC tx is present (https://github.com/dymensionxyz/dymint/issues/709)
 			nextEmptyBlock = time.Now().Add(m.Conf.MaxIdleTime)
 			if 0 < len(block.Data.Txs) {
+				// the block wasn't empty so we want to make sure we don't wait too long before producing another one, in order to facilitate proofs for ibc
+				// TODO: optimize to only do this if IBC transactions are present (https://github.com/dymensionxyz/dymint/issues/709)
 				nextEmptyBlock = time.Now().Add(m.Conf.MaxProofTime)
 			} else {
 				m.logger.Info("Produced empty block.")
 			}
 
-			// Send the size to the accumulated size channel
-			// This will block in case the submitter is too slow and it's buffer is full
-			size := uint64(block.ToProto().Size()) + uint64(commit.ToProto().Size())
+			pause := len(bytesProducedC) == cap(bytesProducedC) // here we assume cap is at least 1
```

Comment:
I think something like this would be more concise and easier to understand:

```
select {
	case <-ctx.Done():
		return nil
	case bytesProducedC <- block.SizeBytes() + commit.SizeBytes():
	default:
		evt := &events.DataHealthStatus{Error: fmt.Errorf("bytes produced channel is full: %w", gerrc.ErrResourceExhausted)}
		uevent.MustPublish(ctx, m.Pubsub, evt, events.HealthStatusList)
		m.logger.Error("Enough bytes to build a batch have been accumulated, but too many batches are pending submission. Pausing block production until a signal is consumed.")

		bytesProducedC <- block.SizeBytes() + commit.SizeBytes()
		evt = &events.DataHealthStatus{Error: nil}
		uevent.MustPublish(ctx, m.Pubsub, evt, events.HealthStatusList)
		m.logger.Info("Resumed block production.")
}
```

Reply:
That looks wrong because you are not waiting on context on the second `bytesProducedC <- block.SizeBytes() + commit.SizeBytes()` line